src/main/scala/firrtl/passes/ConstProp.scala

```diff
@@ -0,0 +1,198 @@
+      case IntWidth(w) if w == 1 && long_BANG(tpe(rhs)) == w =>
+        if (lhs.value == 0) rhs // neq(x, 0) => x
+        else e
+      case _ => e
+    }
+  }
+
+  private def constPropPrim(e: DoPrim): Expression = e.op match {
+    case SHIFT_RIGHT_OP => {
+      val amount = e.consts(0).toInt
+      def shiftWidth(w: Width) = (w - IntWidth(amount)) max IntWidth(1)
+      e.args(0) match {
+        // TODO when amount >= x.width, return a zero-width wire
+        case UIntValue(v, w) => UIntValue(v >> amount, shiftWidth(w))
+        // take sign bit if shift amount is larger than arg width
+        case SIntValue(v, w) => SIntValue(v >> amount, shiftWidth(w))
+        case _ => e
+      }
+    }
+    case AND_OP => FoldAND(e)
+    case OR_OP => FoldOR(e)
+    case XOR_OP => FoldXOR(e)
+    case EQUAL_OP => FoldEqual(e)
+    case NEQUAL_OP => FoldNotEqual(e)
+    case NOT_OP => e.args(0) match {
+      case UIntValue(v, IntWidth(w)) => UIntValue(v ^ ((BigInt(1) << w.toInt) - 1), IntWidth(w))
+      case _ => e
+    }
+    case BITS_SELECT_OP => e.args(0) match {
+      case UIntValue(v, w) => {
+        val hi = e.consts(0).toInt
+        val lo = e.consts(1).toInt
+        require(hi >= lo)
+        UIntValue((v >> lo) & ((BigInt(1) << (hi - lo + 1)) - 1), w)
+      }
+      case x if long_BANG(tpe(e)) == long_BANG(tpe(x)) => tpe(x) match {
+        case t: UIntType => x
+        case _ => DoPrim(AS_UINT_OP, Seq(x), Seq(), tpe(e))
+      }
+      case _ => e
+    }
+    case _ => e
+  }
+
+  private def constPropMuxCond(m: Mux) = (m.cond, tpe(m.tval), tpe(m.fval), m.tpe) match {
+    case (c: UIntValue, ttpe: UIntType, ftpe: UIntType, mtpe: UIntType) =>
+      if (c.value == 1 && ttpe == mtpe) m.tval
+      else if (c.value == 0 && ftpe == mtpe) m.fval
+      else m
+    case _ => m
+  }
+
+  private def constPropMux(m: Mux): Expression = (m.tval, m.fval) match {
+    case (t: UIntValue, f: UIntValue) =>
+      if (t == f) t
+      else if (t.value == 1 && f.value == 0 && long_BANG(m.tpe) == 1) m.cond
+      else constPropMuxCond(m)
+    case _ => constPropMuxCond(m)
+  }
+
+  private def constPropNodeRef(r: WRef, e: Expression) = e match {
+    case _: UIntValue | _: SIntValue | _: WRef => e
+    case _ => r
+  }
+
+  @tailrec
+  private def constPropModule(m: InModule): InModule = {
+    var nPropagated = 0L
+    val nodeMap = collection.mutable.HashMap[String, Expression]()
+
+    def constPropExpression(e: Expression): Expression = {
+      val old = e map constPropExpression
+      val propagated = old match {
+        case p: DoPrim => constPropPrim(p)
+        case m: Mux => constPropMux(m)
+        case r: WRef if nodeMap contains r.name => constPropNodeRef(r, nodeMap(r.name))
+        case x => x
+      }
+      if (old ne propagated)
+        nPropagated += 1
+      propagated
+    }
+
+    def constPropStmt(s: Stmt): Stmt = {
+      s match {
+        case x: DefNode => nodeMap(x.name) = x.value
+        case _ =>
```

Comment:
This is incorrect, constPropStmt should be recursive:
`case _ => s map constPropStmt`